ir/instr.cpp

```diff
@@ -2227,6 +2227,88 @@ unique_ptr<Instr> Memcpy::dup(const string &suffix) const {
 }
 
 
+vector<Value*> Strlen::operands() const {
+  return { ptr };
+}
+
+void Strlen::rauw(const Value &what, Value &with) {
+  RAUW(ptr);
+}
+
+void Strlen::print(ostream &os) const {
+  os << getName() << " = strlen " << *ptr;
+}
+
+
+// Given (bid, ofs0), encode an expression that returns the next offset ofs2
+// that has a null character.
+// Returns: <found?, ofs2>
+static pair<expr, expr>
+nextNull(const Memory &m, const expr &bid, const expr &ofs0, unsigned i,
+         const string &nextnull_uf) {
+  auto ofs = ofs0 + expr::mkUInt(i, ofs0.bits());
+
+  if (i == strlen_unroll_cnt) {
+    auto ofs2 = expr::mkUF(nextnull_uf, {bid, ofs}, ofs);
+    Pointer ptr2(m, bid, ofs2);
+    Byte b = m.load(ptr2);
+    // Approximate the situation by assuming that there is no null between
+    // ofs and ofs2
+    return { ptr2.inbounds(false, true) && !b.is_ptr() && !b.is_poison(false) &&
+             b.is_zero(), move(ofs2) };
+  }
+
+  Pointer ptr(m, bid, ofs);
+  Byte b = m.load(ptr);
+
+  auto is_char = !b.is_ptr() && !b.is_poison(false);
+  auto strictly_inbounds = ptr.inbounds(false, true);
+  if (is_char.isFalse() || strictly_inbounds.isFalse())
+    return { expr(false), ofs };
+  else if (is_char.isTrue() && b.is_zero().isTrue() &&
+           strictly_inbounds.isTrue())
+    return { expr(true), ofs };
+
+  auto [found, ofs_next] = nextNull(m, bid, ofs0, i + 1, nextnull_uf);
+  return { is_char && strictly_inbounds && (b.is_zero() || found),
+           expr::mkIf(b.is_zero(), ofs, ofs_next) };
+}
+
+StateValue Strlen::toSMT(State &s) const {
+  auto &[eptr, np_ptr] = s[*ptr];
+  s.addUB(np_ptr);
+
+  const auto &m = s.getMemory();
+  Pointer p(m, eptr);
+
+  string nextnull_uf = "nextnull_" + m.unique_id();
+  expr bid = expr::mkVar("bid", p.get_bid());
+  expr ofs = expr::mkVar("ofs", p.get_offset());
+  s.addAxiom(
+    expr::mkForAll({ bid, ofs },
+                   expr::mkUF(nextnull_uf, {bid, ofs}, ofs).uge(ofs)));
```

Comment:
given that the definition is now fully unrolled, I think we don't need UFs.
If 2 strlen() calls are made, they will get the same expression anyway.

Reply:
There is the case where the size of a block is unknown, e.g:
```
int f(i8* %ptr) {
return strlen(%ptr);
}
```
In this case, the strlen cannot be fully unrolled, so UF is needed

Reply:
I now see that this depends on whether found should be added to addPre or not

Reply:
Using UFs will lead to false positives, since they don't check equality of memory. E.g. this fails verification:
```
%p = alloca
store %p
return strlen(@glb)
=>
return strlen(@glb)
```

These `strlen` calls will get different UFs, hence verification will fail.
There's no way to make this work beyond what was unrolled.

Reply:
I see. This seems to be a pretty hard problem..
I can discuss this issue with Gil as well. I will share the discussion result.

Reply:
Instead we can make it a random quantified variable (as freeze(undef) returns), which should be not greater than (memblock size - offset). What do you think?

Reply:
I don't think there's any way of completely getting rid of false-positives. Except for the solution I've proposed, which is to assume that the return value of strlen() is between 0 and max_unroll. i.e. ignore programs executions that go beyond this.
Should be fine since strings should be arbitrary, not constant. For constant strings we can constant fold either when converting LLVM IR to Alive, or here by checking if 'load(p[i]) == 0' is a constant bool.